src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/StringSearchValues.cs

```diff
@@ -0,0 +1,365 @@
+                    ? new AsciiStringSearchValuesTeddyNonBucketizedN2<TStartCaseSensitivity, TCaseSensitivity>(values, uniqueValues)
+                    : new AsciiStringSearchValuesTeddyNonBucketizedN3<TStartCaseSensitivity, TCaseSensitivity>(values, uniqueValues);
+            }
+        }
+
+        private static SearchValues<string> CreateForSingleValue(
+            string value,
+            HashSet<string> uniqueValues,
+            bool ignoreCase,
+            bool allAscii,
+            bool asciiLettersOnly)
+        {
+            // We make use of optimizations that may overflow on 32bit systems for long values.
+            int maxLength = IntPtr.Size == 4 ? 1_000_000_000 : int.MaxValue;
+
+            if (Vector128.IsHardwareAccelerated && value.Length > 1 && value.Length <= maxLength)
+            {
+                if (!ignoreCase)
+                {
+                    return new SingleStringSearchValuesThreeChars<CaseSensitive>(value, uniqueValues);
+                }
+
+                if (asciiLettersOnly)
+                {
+                    return new SingleStringSearchValuesThreeChars<CaseInsensitiveAsciiLetters>(value, uniqueValues);
+                }
+
+                if (allAscii)
+                {
+                    return new SingleStringSearchValuesThreeChars<CaseInsensitiveAscii>(value, uniqueValues);
+                }
+
+                // When ignoring casing, all anchor chars we search for must be ASCII.
+                if (char.IsAscii(value[0]) && value.AsSpan().LastIndexOfAnyInRange((char)0, (char)127) > 0)
+                {
+                    return new SingleStringSearchValuesThreeChars<CaseInsensitiveUnicode>(value, uniqueValues);
+                }
+            }
+
+            return ignoreCase
+                ? new SingleStringSearchValuesFallback<SearchValues.TrueConst>(value, uniqueValues)
+                : new SingleStringSearchValuesFallback<SearchValues.FalseConst>(value, uniqueValues);
+        }
+
+        private static void AnalyzeValues(
+            ReadOnlySpan<string> values,
+            ref bool ignoreCase,
+            out bool allAscii,
+            out bool asciiLettersOnly,
+            out bool nonAsciiAffectedByCaseConversion,
+            out int minLength)
+        {
+            allAscii = true;
+            asciiLettersOnly = true;
+            minLength = int.MaxValue;
+
+            foreach (string value in values)
+            {
+                allAscii = allAscii && Ascii.IsValid(value);
+                asciiLettersOnly = asciiLettersOnly && !value.AsSpan().ContainsAnyExcept(s_asciiLetters);
+                minLength = Math.Min(minLength, value.Length);
+            }
+
+            // TODO: Not all characters participate in Unicode case conversion.
+            // If we can determine that none of the non-ASCII characters do, we can make searching faster
+            // by using the same paths as we do for ASCII-only values.
```

Comment:
This TODO is purely a future possible optimization, yes?

Reply:
Yes. The three TODOs left in the code (this, bucket selection strategy, using first N chars for teddy) are things that we can look into improving in the future, but shouldn't block this PR.
I'll reword them to something like "Potential optimization" so it's more clear.